qiskit/providers/basic_provider/basic_provider_tools.py

```diff
@@ -0,0 +1,188 @@
+
+import numpy as np
+
+import qiskit.circuit.library.standard_gates as gates
+from qiskit.exceptions import QiskitError
+
+# Single qubit gates supported by ``single_gate_params``.
+SINGLE_QUBIT_GATES = ("U", "u", "h", "p", "u1", "u2", "u3", "rz", "sx", "x")
+
+
+def single_gate_matrix(gate: str, params: list[float] | None = None) -> np.ndarray:
+    """Get the matrix for a single qubit.
+
+    Args:
+        gate: the single qubit gate name
+        params: the operation parameters op['params']
+    Returns:
+        array: A numpy array representing the matrix
+    Raises:
+        QiskitError: If a gate outside the supported set is passed in for the
+            ``Gate`` argument.
+    """
+    if params is None:
+        params = []
+
+    if gate == "U":
+        gc = gates.UGate
+    elif gate == "u3":
+        gc = gates.U3Gate
+    elif gate == "h":
+        gc = gates.HGate
+    elif gate == "u":
+        gc = gates.UGate
+    elif gate == "p":
+        gc = gates.PhaseGate
+    elif gate == "u2":
+        gc = gates.U2Gate
+    elif gate == "u1":
+        gc = gates.U1Gate
+    elif gate == "rz":
+        gc = gates.RZGate
+    elif gate == "id":
+        gc = gates.IGate
+    elif gate == "sx":
+        gc = gates.SXGate
+    elif gate == "x":
+        gc = gates.XGate
+    else:
+        raise QiskitError("Gate is not a valid basis gate for this simulator: %s" % gate)
+
+    return gc(*params).to_matrix()
+
+
+# Cache CX matrix as no parameters.
+_CX_MATRIX = gates.CXGate().to_matrix()
+
+
+def cx_gate_matrix() -> np.ndarray:
+    """Get the matrix for a controlled-NOT gate."""
+    return np.array([[1, 0, 0, 0], [0, 0, 0, 1], [0, 0, 1, 0], [0, 1, 0, 0]], dtype=complex)
+
+
+def einsum_matmul_index(gate_indices: list[int], number_of_qubits: int) -> str:
+    """Return the index string for Numpy.einsum matrix-matrix multiplication.
+
+    The returned indices are to perform a matrix multiplication A.B where
+    the matrix A is an M-qubit matrix, matrix B is an N-qubit matrix, and
+    M <= N, and identity matrices are implied on the subsystems where A has no
+    support on B.
+
+    Args:
+        gate_indices (list[int]): the indices of the right matrix subsystems
+                                   to contract with the left matrix.
+        number_of_qubits (int): the total number of qubits for the right matrix.
+
+    Returns:
+        str: An indices string for the Numpy.einsum function.
+    """
+
+    mat_l, mat_r, tens_lin, tens_lout = _einsum_matmul_index_helper(gate_indices, number_of_qubits)
+
+    # Right indices for the N-qubit input and output tensor
+    tens_r = ascii_uppercase[:number_of_qubits]
+
+    # Combine indices into matrix multiplication string format
+    # for numpy.einsum function
+    return "{mat_l}{mat_r}, ".format(
+        mat_l=mat_l, mat_r=mat_r
+    ) + "{tens_lin}{tens_r}->{tens_lout}{tens_r}".format(
+        tens_lin=tens_lin, tens_lout=tens_lout, tens_r=tens_r
+    )
```

Comment:
Not that it matters too much, but doing `+` makes 3 string objects and copies between them. We can do this a single f string and avoid that overhead.

```suggestion
    return f"{mat_l}{mat_r}, {tens_lin}{tens_r}->{tens_lout}{tens_r}"
```